Cpp_hands_on_programming/lib/shm/Ipc_shm.cc

```diff
@@ -7,191 +7,201 @@
-
-    // fd = shm_open(name.c_str(), O_RDWR | O_CREAT, 0666);
-	// if (fd == -1) { 
-	// 	throw(std::runtime_error("Ipc_shm::send: shm_open"));
-	// }
-	// // allocate the size of the shared memory object
-	// if (ftruncate(fd, sizeof(Ipc_shm)) == -1) {
-	// 	throw(std::runtime_error("Ipc_shm::send: ftruncate"));	
-	// }
-	// // get a pointer to shared memory region
-	// auto shm_ptr = (Ipc_shm *)mmap(NULL, sizeof(Ipc_shm), PROT_WRITE, MAP_SHARED, fd, 0);
-	// if (shm_ptr == MAP_FAILED) {
-	// 	throw(std::runtime_error("Ipc_shm::send: mmap"));
-	// }
-    // //close fd
-	// close(fd);
+ 	if ((size = getFileSize(filename)) == -1) {
+		throw(std::runtime_error("Ipc_method:: getFileSize"));
+	}
+
+    fd = shm_open(name.c_str(), O_RDWR | O_CREAT, 0666);
+	if (fd == -1) { 
+		throw(std::runtime_error("Ipc_shm::send: shm_open"));
+	}
+	// allocate the size of the shared memory object
+	if (ftruncate(fd, sizeof(Ipc_shm)) == -1) {
+		throw(std::runtime_error("Ipc_shm::send: ftruncate"));	
+	}
+	// get a pointer to shared memory region
+	auto shm_ptr = (Ipc_shm *)mmap(NULL, sizeof(Ipc_shm), PROT_WRITE, MAP_SHARED, fd, 0);
+	if (shm_ptr == MAP_FAILED) {
+		throw(std::runtime_error("Ipc_shm::send: mmap"));
+	}
+    //close fd
+	close(fd);
+
+	shm_ptr->mutex_init();
+	shm_ptr->condvar_init();
+	// initialized shared memory object member values
+	shm_ptr->mutex_lock();
+	shm_ptr->sent = false;
+	shm_ptr->end = false;
+	shm_ptr->init = true;
+	shm_ptr->mutex_unlock();
 	
-	// if ((bytes_read = readFile(filename, total, v, DATA_SIZE) == -1)) {
-	// 	throw(std::runtime_error("Ipc_shm::send: readFile"));
-	// }
-
-	// shm_ptr->mutex_init();
-	// shm_ptr->condvar_init();
-	// // initialized shared memory object member values
-	// shm_ptr->sent = false;
-	// shm_ptr->end = false;
-	// shm_ptr->init = true;
-
-    // while (total < size) {
-    //     // shm_ptr->m.lock();
-	// 	shm_ptr->mutex_lock();
-	// 	// check sent status and blocked if data still has not been retrieved 
-	// 	while (shm_ptr->sent) {
-	// 		shm_ptr->condvar_wait();
-	// 	}
-	// 	//read the file into shared buffer;
-	// 	if ((bytes_read = readFile(filename, total, v, DATA_SIZE)) == -1) {
-	// 		throw(std::runtime_error("Ipc_shm::send: readFile"));
-	// 	}
-	// 	copy(v.begin(), v.end(), shm_ptr->buffer);
-	// 	total += bytes_read;
-	// 	if (total == size) {
-	// 		shm_ptr->end = true;
-	// 	}
-	// 	// update shared memory object
-	// 	shm_ptr->sent = true;
-	// 	shm_ptr->data_size = bytes_read; 
-
-	// 	shm_ptr->mutex_unlock();
-	// 	//wake up process which is currently condvar blocked
-	// 	shm_ptr->condvar_broadcast();
-	// }
-
-	// if (total == size) {
-	// 	std::cout << "File delivered by shm successfully, exiting the program.."  << std::endl;
-	// } else {
-	// 	throw(std::runtime_error("Ipc_shm::send: total and size difference"));
-	// }
+	// wait for receiver to connect the shared memory region
+	while (shm_ptr->init) {  
+		std::this_thread::sleep_for(std::chrono::milliseconds(1));
```

Comment:
> If you want to improve or change to more C++ way, you can check about std::mutex instead of using C mutexes.

I used it initially, unfortunately std::mutex is working only in multi threading environment, it does not support IPC :)